state/storage.go

```diff
@@ -938,3 +936,27 @@ func (s *Storage) Teardown() {
 		panic("Storage.Teardown: " + err.Error())
 	}
 }
+
+// circularSlice is a slice which can be appended to which will wraparound at `max`.
+// Mostly useful for lazily calculating heroes. The values returned aren't sorted.
+type circularSlice struct {
+	i    int
+	vals []int64
+	max  int
+}
```

Comment:
Put differently: it is a gadget into which you feed a sequence of values, outputting the `max` most recent values at the end.